drivers/can/can_socketcan.c

```diff
@@ -0,0 +1,161 @@
+#include <devicetree.h>
+#include <device.h>
+
+#include <logging/log.h>
+LOG_MODULE_REGISTER(net_socketcan, CONFIG_CAN_LOG_LEVEL);
+
+#define SEND_TIMEOUT K_MSEC(100)
+
+struct socket_can_context {
+	struct net_if *iface;
+};
+
+struct socket_can_config {
+	const struct device *can_dev;
+};
+
+static void socket_can_recv(struct zcan_frame *frame, void *arg)
+{
+	struct socket_can_context *ctx = (struct socket_can_context *)arg;
+	struct net_pkt *pkt;
+	int ret;
+
+	LOG_DBG("pkt on interface %p", ctx->iface);
+	pkt = net_pkt_rx_alloc_with_buffer(ctx->iface, sizeof(struct zcan_frame),
+					   AF_CAN, 0, K_NO_WAIT);
+	if (!pkt) {
+		LOG_ERR("Failed to obtain net_pkt");
+		return;
+	}
+
+	if (net_pkt_write(pkt, frame, sizeof(struct zcan_frame))) {
+		LOG_ERR("Failed to append RX data");
+		net_pkt_unref(pkt);
+		return;
+	}
+
+	ret = net_recv_data(ctx->iface, pkt);
+	if (ret < 0) {
+		LOG_DBG("net_recv_data failed [%d]", ret);
+		net_pkt_unref(pkt);
+	}
+}
+
+static int socket_can_setsockopt(const struct device *dev, void *obj,
+					int level, int optname,
+					const void *optval, socklen_t optlen)
+{
+	const struct socket_can_config *cfg = dev->config;
+	struct socket_can_context *socket_context = dev->data;
+	struct net_context *ctx = obj;
+	int ret;
+
+	if (level != SOL_CAN_RAW && optname != CAN_RAW_FILTER) {
+		errno = EINVAL;
+		return -1;
+	}
+
+	__ASSERT_NO_MSG(optlen == sizeof(struct zcan_filter));
+
+	ret = can_add_rx_filter(cfg->can_dev, socket_can_recv, socket_context, optval);
+	if (ret == -ENOSPC) {
+		errno = ENOSPC;
+		return -1;
+	}
+
+	net_context_set_filter_id(ctx, ret);
+
+	return 0;
+}
+
+static void socket_can_close(const struct device *dev, int filter_id)
+{
+	const struct socket_can_config *cfg = dev->config;
+
+	can_remove_rx_filter(cfg->can_dev, filter_id);
+}
+
+static void socket_can_send_tx_callback(int error, void *arg)
+{
+	if (error != 0) {
+		LOG_DBG("socket CAN TX error [%d]", error);
+	}
+}
+
+static int socket_can_send(const struct device *dev,
+				  struct net_pkt *pkt)
+{
+	const struct socket_can_config *cfg = dev->config;
+	int ret;
+
+	if (net_pkt_family(pkt) != AF_CAN) {
+		return -EPFNOSUPPORT;
+	}
+
+	ret = can_send(cfg->can_dev,
+		       (struct zcan_frame *)pkt->frags->data,
+		       SEND_TIMEOUT, socket_can_send_tx_callback, NULL);
+
+	if (ret == 0) {
+		net_pkt_unref(pkt);
+	} else {
+		LOG_DBG("Cannot send socket CAN msg (%d)", ret);
+	}
+
+	/* If something went wrong, then we need to return negative value to
+	 * net_if.c:net_if_tx() so that the net_pkt will get released.
+	 */
+	return ret;
+}
+
+static void socket_can_iface_init(struct net_if *iface)
+{
+	const struct device *dev = net_if_get_device(iface);
+	struct socket_can_context *socket_context = dev->data;
+
+	socket_context->iface = iface;
+
+	LOG_DBG("Init CAN interface %p dev %p", iface, dev);
+}
+
+static int socket_can_init(const struct device *dev)
+{
+	const struct socket_can_config *cfg = dev->config;
+
+	ARG_UNUSED(cfg);
```

Comment:
`cfg` is actually used in the `LOG_DBG()` macro call.

Reply:
But only if DBG is built.

Reply:
Still used, though.